File: frontend/src/technicals.ts
import type { PricePoint } from "./api/client";

export type Signal = "bullish" | "bearish" | "neutral";

function sma(values: (number | null)[], period: number): (number | null)[] {
  const out: (number | null)[] = new Array(values.length).fill(null);
  let sum = 0;
  let count = 0;
  for (let i = 0; i < values.length; i++) {
    const v = values[i];
    if (v !== null) {
      sum += v;
      count++;
    }
    if (i >= period) {
      const drop = values[i - period];
      if (drop !== null) {
        sum -= drop;
        count--;
      }
    }
    out[i] = i >= period - 1 && count === period ? sum / period : null;
  }
  return out;
}

// Wilder's RSI (the standard 14-period formula used by most charting platforms).
function rsi(closes: number[], period = 14): (number | null)[] {
  const out: (number | null)[] = new Array(closes.length).fill(null);
  if (closes.length < period + 1) return out;

  let avgGain = 0;
  let avgLoss = 0;
  for (let i = 1; i <= period; i++) {
    const change = closes[i] - closes[i - 1];
    avgGain += Math.max(change, 0);
    avgLoss += Math.max(-change, 0);
  }
  avgGain /= period;
  avgLoss /= period;
  out[period] = avgLoss === 0 ? 100 : 100 - 100 / (1 + avgGain / avgLoss);

  for (let i = period + 1; i < closes.length; i++) {
    const change = closes[i] - closes[i - 1];
    const gain = Math.max(change, 0);
    const loss = Math.max(-change, 0);
    avgGain = (avgGain * (period - 1) + gain) / period;
    avgLoss = (avgLoss * (period - 1) + loss) / period;
    out[i] = avgLoss === 0 ? 100 : 100 - 100 / (1 + avgGain / avgLoss);
  }
  return out;
}

function averageTrueRangePct(points: PricePoint[], period = 14): number | null {
  if (points.length < period + 1) return null;
  const trueRanges: number[] = [];
  for (let i = 1; i < points.length; i++) {
    const { high, low } = points[i];
    const prevClose = points[i - 1].close;
    if (high === null || low === null) continue;
    trueRanges.push(Math.max(high - low, Math.abs(high - prevClose), Math.abs(low - prevClose)));
  }
  const recent = trueRanges.slice(-period);
  if (recent.length < period) return null;
  const atr = recent.reduce((a, b) => a + b, 0) / recent.length;
  return atr / points[points.length - 1].close;
}

export interface TechnicalRead {
  latestClose: number;
  sma20: number | null;
  sma50: number | null;
  sma200: number | null;
  rsi14: number | null;
  trend: { signal: Signal; label: string };
  momentum: { signal: Signal; label: string };
  volume: { signal: Signal; label: string; ratio: number | null };
  range52w: { high: number; low: number; pctFromHigh: number; pctFromLow: number };
  nearTerm: { support: number; resistance: number };
  volatility: { atrPct: number | null; label: string };
  chartSeries: { date: string; close: number; sma50: number | null; sma200: number | null }[];
}

export function computeTechnicalRead(points: PricePoint[]): TechnicalRead | null {
  if (points.length < 30) return null;

  const closes = points.map((p) => p.close);
  const sma20Arr = sma(closes, 20);
  const sma50Arr = sma(closes, 50);
  const sma200Arr = sma(closes, 200);
  const rsiArr = rsi(closes, 14);

  const last = closes.length - 1;
  const latestClose = closes[last];
  const sma20v = sma20Arr[last];
  const sma50v = sma50Arr[last];
  const sma200v = sma200Arr[last];
  const rsiV = rsiArr[last];

  // Trend: classic moving-average structure a technical trader checks first.
  let trendSignal: Signal = "neutral";
  let trendLabel = "Not enough history to assess trend (need 200+ trading days).";
  if (sma50v !== null && sma200v !== null) {
    const aboveBoth = latestClose > sma50v && latestClose > sma200v;
    const belowBoth = latestClose < sma50v && latestClose < sma200v;
    const goldenCross = sma50v > sma200v;
    if (aboveBoth && goldenCross) {
      trendSignal = "bullish";
      trendLabel = "Uptrend: price is above both the 50- and 200-day moving averages, and the 50-day is above the 200-day (golden cross) — bullish trend structure.";
    } else if (belowBoth && !goldenCross) {
      trendSignal = "bearish";
      trendLabel = "Downtrend: price is below both moving averages, and the 50-day is below the 200-day (death cross) — bearish trend structure.";
    } else {
      trendSignal = "neutral";
      trendLabel = `Mixed trend: price and moving averages disagree (${goldenCross ? "50-day above" : "50-day below"} 200-day) — no clean directional structure right now.`;
    }
  }

  // Momentum: RSI(14), the standard overbought/oversold oscillator.
  let momentumSignal: Signal = "neutral";
  let momentumLabel = "RSI unavailable (need 15+ trading days).";
  if (rsiV !== null) {
    if (rsiV >= 70) {
      momentumSignal = "bearish";
      momentumLabel = `RSI ${rsiV.toFixed(0)} — overbought (>70). Momentum has run hot; often precedes a pullback or consolidation, though strong trends can stay overbought a while.`;
    } else if (rsiV <= 30) {
      momentumSignal = "bullish";
      momentumLabel = `RSI ${rsiV.toFixed(0)} — oversold (<30). Selling has been aggressive; often precedes a bounce, though downtrends can stay oversold a while.`;
    } else {
      momentumSignal = "neutral";
      momentumLabel = `RSI ${rsiV.toFixed(0)} — neutral range (30–70), no momentum extreme.`;
    }
  }

  // Volume: is recent participation confirming the move, or is the move on thin volume?
  const volumes = points.map((p) => p.volume ?? 0);
  const recentVol = volumes.slice(-10);
  const baselineVol = volumes.slice(-60, -10);
  const recentAvg = recentVol.reduce((a, b) => a + b, 0) / (recentVol.length || 1);
  const baselineAvg = baselineVol.length ? baselineVol.reduce((a, b) => a + b, 0) / baselineVol.length : null;
  const volRatio = baselineAvg && baselineAvg > 0 ? recentAvg / baselineAvg : null;
  let volumeSignal: Signal = "neutral";
  let volumeLabel = "Not enough volume history to compare.";
  if (volRatio !== null) {
    if (volRatio >= 1.3) {
      volumeSignal = trendSignal === "bearish" ? "bearish" : "bullish";
      volumeLabel = `10-day average volume is ${volRatio.toFixed(1)}x the prior 50-day average — elevated participation, which tends to confirm the current move rather than a low-conviction drift.`;
    } else if (volRatio <= 0.7) {
      volumeSignal = "neutral";
      volumeLabel = `10-day average volume is only ${volRatio.toFixed(1)}x the prior 50-day average — the recent move is happening on light volume, which is a weaker signal.`;
    } else {
      volumeSignal = "neutral";
      volumeLabel = `10-day average volume is in line with the prior 50-day average (${volRatio.toFixed(1)}x) — no unusual participation either way.`;
    }
  }

  // Position in the 52-week range (from actual daily highs/lows, not just closes).
  const highs = points.map((p) => p.high ?? p.close);
  const lows = points.map((p) => p.low ?? p.close);
  const window52w = Math.min(points.length, 252);
  const high52w = Math.max(...highs.slice(-window52w));
  const low52w = Math.min(...lows.slice(-window52w));
  const pctFromHigh = (latestClose - high52w) / high52w;
  const pctFromLow = (latestClose - low52w) / low52w;

  // Near-term support/resistance: simple 20-trading-day (~1 month) swing range.
  const window20 = Math.min(points.length, 20);
  const support = Math.min(...lows.slice(-window20));
  const resistance = Math.max(...highs.slice(-window20));

  const atrPct = averageTrueRangePct(points, 14);
  let volatilityLabel = "Not enough data to compute ATR.";
  if (atrPct !== null) {
    const level = atrPct > 0.035 ? "high" : atrPct < 0.015 ? "low" : "moderate";
    volatilityLabel = `Average true range is ${(atrPct * 100).toFixed(1)}% of price per day (${level} volatility) — sets a rough expectation for normal day-to-day swings.`;
  }

  const chartSeries = points.map((p, i) => ({
    date: p.date,
    close: p.close,
    sma50: sma50Arr[i],
    sma200: sma200Arr[i],
  }));

  return {
    latestClose,
    sma20: sma20v,
    sma50: sma50v,
    sma200: sma200v,
    rsi14: rsiV,
    trend: { signal: trendSignal, label: trendLabel },
    momentum: { signal: momentumSignal, label: momentumLabel },
    volume: { signal: volumeSignal, label: volumeLabel, ratio: volRatio },
    range52w: { high: high52w, low: low52w, pctFromHigh, pctFromLow },
    nearTerm: { support, resistance },
    volatility: { atrPct, label: volatilityLabel },
    chartSeries,
  };
}

File: backend/app/signals/technicals.py
"""Python port of frontend/src/technicals.ts, for the daily scanner.

The formulas and thresholds are deliberately identical to the TypeScript
version that powers the Technical Read panel -- same SMA, same Wilder RSI,
same ATR -- so the scanner can never disagree with what the UI shows for
the same ticker on the same day.

The one deliberate difference: the UI only needs *today's* values to
describe current state, while the scanner needs **yesterday's too**, so it
can fire on a *crossing* (sma50 crossed sma200 today) rather than a state
(sma50 is above sma200, which would alert every day forever). Everything
here therefore returns series or explicit prev/current pairs rather than
collapsing to a single reading.
"""

from __future__ import annotations

from dataclasses import dataclass


def sma(values: list[float | None], period: int) -> list[float | None]:
    """Simple moving average; None until a full window of real values."""
    out: list[float | None] = [None] * len(values)
    total = 0.0
    count = 0
    for i, v in enumerate(values):
        if v is not None:
            total += v
            count += 1
        if i >= period:
            drop = values[i - period]
            if drop is not None:
                total -= drop
                count -= 1
        out[i] = (total / period) if (i >= period - 1 and count == period) else None
    return out


def rsi(closes: list[float], period: int = 14) -> list[float | None]:
    """Wilder's RSI -- the standard 14-period formula charting platforms use."""
    out: list[float | None] = [None] * len(closes)
    if len(closes) < period + 1:
        return out

    avg_gain = 0.0
    avg_loss = 0.0
    for i in range(1, period + 1):
        change = closes[i] - closes[i - 1]
        avg_gain += max(change, 0.0)
        avg_loss += max(-change, 0.0)
    avg_gain /= period
    avg_loss /= period
    out[period] = 100.0 if avg_loss == 0 else 100.0 - 100.0 / (1 + avg_gain / avg_loss)

    for i in range(period + 1, len(closes)):
        change = closes[i] - closes[i - 1]
        avg_gain = (avg_gain * (period - 1) + max(change, 0.0)) / period
        avg_loss = (avg_loss * (period - 1) + max(-change, 0.0)) / period
        out[i] = 100.0 if avg_loss == 0 else 100.0 - 100.0 / (1 + avg_gain / avg_loss)
    return out


def average_true_range_pct(points: list[dict], period: int = 14) -> float | None:
    if len(points) < period + 1:
        return None
    true_ranges: list[float] = []
    for i in range(1, len(points)):
        high, low = points[i].get("high"), points[i].get("low")
        prev_close = points[i - 1]["close"]
        if high is None or low is None:
            continue
        true_ranges.append(max(high - low, abs(high - prev_close), abs(low - prev_close)))
    recent = true_ranges[-period:]
    if len(recent) < period:
        return None
    return (sum(recent) / len(recent)) / points[-1]["close"]


@dataclass
class TechnicalRead:
    latest_close: float
    prev_close: float | None
    sma50: float | None
    sma50_prev: float | None
    sma200: float | None
    sma200_prev: float | None
    rsi14: float | None
    rsi14_prev: float | None
    volume_ratio: float | None
    # 52-week extremes computed EXCLUDING today, so "today broke the high"
    # is a real breakout rather than trivially true (today's own high is
    # always part of a window that includes today).
    high_52w_prior: float | None
    low_52w_prior: float | None
    atr_pct: float | None


def compute_technical_read(points: list[dict]) -> TechnicalRead | None:
    """`points` are price bars from /api/companies/{ticker}/prices, oldest first."""
    if len(points) < 30:
        return None

    closes = [p["close"] for p in points]
    sma50_arr = sma(list(closes), 50)
    sma200_arr = sma(list(closes), 200)
    rsi_arr = rsi(closes, 14)

    def at(arr: list, idx: int):
        return arr[idx] if len(arr) >= abs(idx) else None

    volumes = [p.get("volume") or 0 for p in points]
    recent_vol = volumes[-10:]
    baseline_vol = volumes[-60:-10]
    recent_avg = sum(recent_vol) / len(recent_vol) if recent_vol else 0.0
    baseline_avg = (sum(baseline_vol) / len(baseline_vol)) if baseline_vol else None
    vol_ratio = (recent_avg / baseline_avg) if (baseline_avg and baseline_avg > 0) else None

    highs = [(p.get("high") if p.get("high") is not None else p["close"]) for p in points]
    lows = [(p.get("low") if p.get("low") is not None else p["close"]) for p in points]
    # Exclude the final bar: the window is the *prior* 52 weeks.
    prior_highs = highs[-253:-1]
    prior_lows = lows[-253:-1]

    return TechnicalRead(
        latest_close=closes[-1],
        prev_close=at(closes, -2),
        sma50=at(sma50_arr, -1),
        sma50_prev=at(sma50_arr, -2),
        sma200=at(sma200_arr, -1),
        sma200_prev=at(sma200_arr, -2),
        rsi14=at(rsi_arr, -1),
        rsi14_prev=at(rsi_arr, -2),
        volume_ratio=vol_ratio,
        high_52w_prior=max(prior_highs) if prior_highs else None,
        low_52w_prior=min(prior_lows) if prior_lows else None,
        atr_pct=average_true_range_pct(points, 14),
    )
